hibernate-envers/src/main/java/org/hibernate/envers/internal/entities/mapper/relation/ToOneIdMapper.java

```diff
@@ -49,10 +54,23 @@ public boolean mapToMapFromEntity(
 			Object oldObj) {
 		final HashMap<String, Object> newData = new HashMap<>();
 
+		Object oldObject = oldObj;
+		Object newObject = newObj;
+		if ( lazyMapping ) {
+			if ( nonInsertableFake && oldObject instanceof HibernateProxy ) {
+				oldObject = ( (HibernateProxy) oldObject ).getHibernateLazyInitializer().getImplementation();
+			}
+
+
+			if ( !nonInsertableFake && newObject instanceof HibernateProxy ) {
+				newObject = ( (HibernateProxy) newObject ).getHibernateLazyInitializer().getImplementation();
+			}
+		}
+
 		// If this property is originally non-insertable, but made insertable because it is in a many-to-one "fake"
 		// bi-directional relation, we always store the "old", unchaged data, to prevent storing changes made
 		// to this field. It is the responsibility of the collection to properly update it if it really changed.
-		delegate.mapToMapFromEntity( newData, nonInsertableFake ? oldObj : newObj );
+		delegate.mapToMapFromEntity( newData, nonInsertableFake ? oldObject : newObject );
```

Comment:
If I didn't misunderstand, your logic seems as straightforward as follows:
`
Object obj = nonInsertableFake ? oldObj : newObj;
if ( lazyMapping && obj instanceof HibernateProxy ) {
    obj = ( (HibernateProxy) obj).getHibernateLazyInitializer().getImplementation( session );
}
delegate.mapToMapFromEntity( newDate, obj );
`
I do think the above code is easier to maintain because it involves no logic duplication and easier to understand.
IMHO.

Reply:
Nope, I completely agree, thanks for the suggestion!